### PR TITLE
Implement identity in PostgresAdapter

### PR DESCRIPTION
**Pull request by ajibarra**
Working on #2070 . Implementing IDENTITY for column in PostgresAdapter.

### CODE REVIEW

**Comment by swiffer** [comment]
@MasterOdin any Chance to get this merged for 0.13? Would be great to have Support in cakephp/Migrations at some point. Thanks a Lot!

File: src/Phinx/Db/Adapter/PostgresAdapter.php [hunk 1]
```diff
@@ -393,15 +416,14 @@ public function getColumns(string $tableName): array
         $sql = sprintf(
             'SELECT column_name, data_type, udt_name, is_identity, is_nullable,
              column_default, character_maximum_length, numeric_precision, numeric_scale,
-             datetime_precision
+             datetime_precision, identity_generation
```
Comment (MasterOdin) [inline src/Phinx/Db/Adapter/PostgresAdapter.php]: Easy enough I think to change this to:

```php
        $sql = sprintf(
            'SELECT column_name, data_type, udt_name, is_identity, is_nullable,
             column_default, character_maximum_length, numeric_precision, numeric_scale,
             datetime_precision
             %s
             FROM information_schema.columns
             WHERE table_schema = %s AND table_name = %s
             ORDER BY ordinal_position',
            $this->useIdentity ? ', identity_generation' : '',
            $this->getConnection()->quote($parts['schema']),
            $this->getConnection()->quote($parts['table'])
```

and maintain compatibility with any and all PG versions (mostly a thought for redshift)

File: src/Phinx/Db/Adapter/PostgresAdapter.php [hunk 2]
```diff
@@ -1156,7 +1222,12 @@ protected function getDefaultValueDefinition($default, ?string $columnType = nul
     protected function getColumnSqlDefinition(Column $column): string
     {
         $buffer = [];
-        if ($column->isIdentity()) {
+
+//        var_dump($column->getName());
+//        var_dump($column->isIdentity());
+//        var_dump($this->useIdentity);
+//        var_dump($column->getGenerated());
```
Comment (MasterOdin) [inline src/Phinx/Db/Adapter/PostgresAdapter.php]: remove?

File: src/Phinx/Db/Adapter/PostgresAdapter.php [hunk 3]
```diff
@@ -42,6 +45,21 @@ class PostgresAdapter extends PdoAdapter
...
+     */
+    protected $useIdentity;
+
+    /**
+     * @return bool
+     */
+    public function isUseIdentity(): bool
+    {
+        return $this->useIdentity;
+    }
```
Comment (MasterOdin) [inline src/Phinx/Db/Adapter/PostgresAdapter.php]: As opposed to having this public function just for use in tests, I'd rather we do the same as we do for mysql, and just add an equivalent function to the test suite and use that:

https://github.com/cakephp/phinx/blob/9a6ce1e7fdf0fa4e602ba5875b5bc9442ccaa115/tests/Phinx/Db/Adapter/MysqlAdapterTest.php#L45-L48

It's a bit repetitive, but I'd rather avoid adding new public functions, especially ones that are very adapter specific. Just another thing that gets added for BC, and feels like easier to potentially break.

File: tests/Phinx/Db/Adapter/PostgresAdapterTest.php [hunk 4]
```diff
@@ -832,6 +902,80 @@ public function testChangeColumn()
...
+        $table->changeColumn('column1', 'integer', ['identity' => true, 'generated' => PostgresAdapter::GENERATED_BY_DEFAULT]);
+        $table->save();
+        $columns = $this->adapter->getColumns('table1');
+        foreach ($columns as $column) {
+            if ($column->getName() === 'column1') {
+                $this->assertTrue($column->getIdentity());
+                $this->assertEquals(PostgresAdapter::GENERATED_BY_DEFAULT, $column->getGenerated());
+            }
+        }
+    }
```
Comment (MasterOdin) [inline tests/Phinx/Db/Adapter/PostgresAdapterTest.php:919]: Can we combine these two tests into one and use a dataProvider?

File: tests/Phinx/Db/Adapter/PostgresAdapterTest.php [hunk 5]
```diff
@@ -537,6 +537,76 @@ public function testAddColumnWithDefaultZero()
...
+        $table->addColumn('id', 'integer', ['identity' => true, 'generated' => null])
+            ->save();
+        $columns = $this->adapter->getColumns('table1');
+        foreach ($columns as $column) {
+            if ($column->getName() === 'id') {
+                $this->assertFalse($column->getIdentity());
+                $this->assertNull($column->getGenerated());
+            }
+        }
+    }
```
Comment (MasterOdin) [inline tests/Phinx/Db/Adapter/PostgresAdapterTest.php]: Same here.

**Review by MasterOdin — CHANGES_REQUESTED** [review]

**Comment by dereuromark** [comment]
ping @ajibarra 

Otherwise @swiffer you could also make a new PR from those changes and include the mentioned feedback
Then we can merge an alternative PR here as well.

File: src/Phinx/Db/Adapter/PostgresAdapter.php [hunk 6]
```diff
@@ -1092,7 +1162,9 @@ public function getPhinxType(string $sqlType): string
             case 'macaddr':
                 return static::PHINX_TYPE_MACADDR;
             default:
-                throw new UnsupportedColumnTypeException('Column type "' . $sqlType . '" is not supported by Postgresql.');
+                throw new UnsupportedColumnTypeException(
+                    'Column type "' . $sqlType . '" is not supported by Postgresql.'
```
Comment (dereuromark) [inline src/Phinx/Db/Adapter/PostgresAdapter.php]: can we use ` for code fencing?

Comment (ajibarra) [reply src/Phinx/Db/Adapter/PostgresAdapter.php]: sure

**Comment by ajibarra** [comment]
@MasterOdin changes requested has been pushed. For some reason coverage has decreased. One of the issues reported is because of splitting the exception message to avoid PHP CS warning but I can't find the actual reason for the other one (L105-L106) because connect method is called for sure and that lines are executed so I am out of ideas.

**Comment by ajibarra** [comment]
> 

@dereuromark Maybe you know?

**Comment by ajibarra** [comment]
> @MasterOdin changes requested has been pushed. For some reason coverage has decreased. One of the issues reported is because of splitting the exception message to avoid PHP CS warning but I can't find the actual reason for the other one (L105-L106) because connect method is called for sure and that lines are executed so I am out of ideas.

@dereuromark Maybe you know?

**Comment by MasterOdin** [comment]
The decrease in code coverage is fine. Most likely some of the formatting changes made here touch upon code that's not tested.

**Comment by ajibarra** [comment]
> The decrease in code coverage is fine. Most likely some of the formatting changes made here touch upon code that's not tested.

Great then, so it is ready to be merged after approval. Thank you @MasterOdin

**Review by MasterOdin — APPROVED** [review]

File: src/Phinx/Db/Adapter/PostgresAdapter.php [hunk 7]
```diff
@@ -92,6 +102,8 @@ public function connect(): void
                 );
             }
 
+            $this->useIdentity = (float)$db->getAttribute(PDO::ATTR_SERVER_VERSION) >= 10;
```
Comment (romanzks) [inline src/Phinx/Db/Adapter/PostgresAdapter.php:105]: I'm not sure if it's the right approach to decide whether to use an `identity` depending on the version of PostgreSQL. This makes it impossible to use the `serial` type to create auto-increment column. These are different things that work in different ways. And one of them does not replace the other. There must be a choice of which auto-increment approach will be used.

**Pull request merged**